### PR TITLE
feat: add UrlSigner interface for customizable URL signing

### PR DESCRIPTION
**Pull request by uriahcarpenter**
## Summary
- Adds generic `UrlSigner` functional interface for custom URL signing during `toString()`
- Refactors `CloudfrontUrlBuilder` to use the new signing mechanism internally
- Adds `expireAt(Instant)` methods to both `CloudfrontUrlBuilder` and `S3UrlBuilder`
- Modernizes internal `ExpireDateHolder` to use `Instant` instead of `Date`
- Adds comprehensive JavaDoc and unit tests
## Test Plan
- [x] All 191 tests pass
- [x] Edge case tests for signing (14 tests)
- [x] Integration tests (5 tests)
- [x] CloudFront expireIn tests (12 tests)

### CODE REVIEW

File: src/test/java/com/widen/urlbuilder/examples/HmacSigningExampleTest.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    
+    // Helper method for HMAC-SHA256
+    private static String hmacSha256(String data, String key) {
+        try {
+            Mac mac = Mac.getInstance("HmacSHA256");
+            SecretKeySpec secretKey = new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
+            mac.init(secretKey);
+            byte[] hash = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
+            return Base64.getEncoder().encodeToString(hash);
```
[the remark this answers is not in the document]
Comment (uriahcarpenter) [reply src/test/java/com/widen/urlbuilder/examples/HmacSigningExampleTest.java]: Updated example with safe base64 generator.

File: src/main/java/com/widen/urlbuilder/CloudfrontUrlBuilder.java [hunk 2]
```diff
@@ -103,16 +209,60 @@ public CloudfrontUrlBuilder expireIn(long duration, TimeUnit unit)
...
+     * @return This builder for method chaining
+     * @deprecated Use {@link #expireAt(Instant)} instead. This method will be removed in a future version.
      */
+    @Deprecated
     public CloudfrontUrlBuilder expireAt(Date date)
     {
-        expireDate.instant = date;
+       InternalUtils.checkNotNull(date, "date");
+       expireDate.instant = date.toInstant();
+       return this;
```
Comment (sagebind) [inline src/main/java/com/widen/urlbuilder/CloudfrontUrlBuilder.java]: Nit: Why 7-space indent?

File: src/main/java/com/widen/urlbuilder/SigningContextImpl.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    private final String url;
+    private final boolean ssl;
+    private final UrlBuilder.GenerationMode generationMode;
+    
+    /**
+     * Creates a new signing context with all fields.
+     * 
+     * @param protocol The protocol (http/https) or empty string for protocol-relative
+     * @param hostname The hostname
+     * @param port The port number or -1 for default
```
Comment (sagebind) [inline src/main/java/com/widen/urlbuilder/SigningContextImpl.java]: Nit: -1 is not a great choice for "unknown" or "default" port. I would use `Integer` instead and allow `null` to be a more explicit intent for default.

File: src/main/java/com/widen/urlbuilder/UrlSafeBase64.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * Encodes the given byte array to a URL-safe Base64 string without padding.
+     *
+     * @param data the bytes to encode
+     * @return URL-safe Base64 encoded string
+     */
+    public static String encode(byte[] data)
+    {
+        return Base64.getUrlEncoder().withoutPadding().encodeToString(data);
```
Comment (sagebind) [inline src/main/java/com/widen/urlbuilder/UrlSafeBase64.java]: `withoutPadding` returns a new `Encoder` instance every call. It would be better to hold this `Encoder` in a private static field and re-use the same `Encoder` instance for each call.
